[content-visibility] Implement initial visibility determination

rwlbuis opened the pull request:
#### b8c6566d556809935878157344ab89bedf59279d
<pre>
[content-visibility] Implement initial visibility determination
<a href="https://bugs.webkit.org/show_bug.cgi?id=259825">https://bugs.webkit.org/show_bug.cgi?id=259825</a>

Reviewed by Tim Nguyen.

Implement initial visibility determination as specified in [1].
This replaces the custom onscreen bookkeeping with the specified
proximity to viewport concept. The updateIntersectionObservations
logic is refactored a bit so it can be used for updating any given
collection of intersection observers.

This PR also changed behaviour to not dispatch ContentVisibilityAutoStateChange
on disconnected elements, this problem is now exposed with the different timing
of initial visibility determination.

[1] <a href="https://github.com/whatwg/html/pull/9312">https://github.com/whatwg/html/pull/9312</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-first-observation-immediate-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-first-observation-immediate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-012-expected.txt:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::unobserve):
(WebCore::ContentVisibilityDocumentState::checkRelevancyOfContentVisibilityElement const): extract per element content relevancy check logic
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements const): bail out early if no update is specified
(WebCore::ContentVisibilityDocumentState::determineInitialVisibleContentVisibility const):
(WebCore::ContentVisibilityDocumentState::updateContentRelevancyStatusForScrollIfNeeded):
(WebCore::ContentVisibilityDocumentState::updateViewportProximity): keep track of per element viewport proximity
(WebCore::ContentVisibilityDocumentState::removeViewportProximity):
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements): Deleted.
(WebCore::ContentVisibilityDocumentState::updateOnScreenObservationTarget): Deleted.
* Source/WebCore/dom/ContentVisibilityDocumentState.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateIntersectionObservations): allow specifying the intersection observers
(WebCore::Document::updateResizeObservations): integrate according to [1]
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::contentVisibilityViewportChange): Deleted.
* Source/WebCore/dom/Element.h:

Canonical link: <a href="https://commits.webkit.org/267779@main">https://commits.webkit.org/267779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74d4bcccbf1e7dedaf5be94fa15321ec1577f782

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16957 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18741 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15875 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17419 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18107 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17157 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17508 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14683 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19554 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14751 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15378 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22105 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15735 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15545 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19867 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16154 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13685 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15316 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4209 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19681 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15985 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.h:
```diff
@@ -26,14 +26,18 @@
 
 enum class DidUpdateAnyContentRelevancy : bool { No, Yes };
+// https://drafts.csswg.org/css-contain/#proximity-to-the-viewport
+enum class ViewportProximity : uint8_t { Far, Near };
```
Not a fan of how "near" can also mean "visible".

rwlbuis replied:
I see what you mean. Something like NearOrIn or InOrNear is more accurate but a bit verbose...

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.h:
```diff
@@ -26,14 +26,18 @@
+// https://drafts.csswg.org/css-contain/#proximity-to-the-viewport
+enum class ViewportProximity : uint8_t { Far, Near };
+
+enum class HadInitialVisibleContentVisibilityDetermination { Yes, No };
```
`bool`, flip the values.

rwlbuis replied:
Done.

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -96,50 +94,82 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
 }
 
-DidUpdateAnyContentRelevancy ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancy>& relevancyToCheck)
+bool ContentVisibilityDocumentState::checkRelevancyOfContentVisibilityElement(Element& target, const OptionSet<ContentRelevancy>& relevancyToCheck) const
```
`OptionSet<>` is small enough to be passed by value.

rwlbuis replied:
Fixed.

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -96,50 +94,82 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
+            newRelevancy.remove(reason);
+    };
+    if (relevancyToCheck.contains(ContentRelevancy::OnScreen)) {
+        auto viewportProximity = m_viewportProximity.find(target);
```
the return value of `find` is an iterator, not viewportProximity itself.

rwlbuis replied:
Yeah, that is unclear indeed, fixed.

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -96,50 +94,82 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
+    return true;
+}
+
+DidUpdateAnyContentRelevancy ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancy>& relevancyToCheck) const
```
Pass OptionSet by value.

rwlbuis replied:
Done.

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -96,50 +94,82 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
+    }
+    auto hadInitialVisibleContentVisibilityDetermination = HadInitialVisibleContentVisibilityDetermination::No;
+    if (!elementsToCheck.isEmpty()) {
+        elementsToCheck.first()->document().updateIntersectionObservations({ m_observer });
```
Is it OK that this happens at a different point in the rendering update than other Intersection Observations? Could state have changed in between?

rwlbuis replied:
The first usage of this intersection observer is special and meant to be same-frame as the element became c-v: auto, specifically to get correct behavior for the resize observers after that it acts just like any other intersection observer. Are you thinking of some specific state? The tests show up no problems.

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -163,21 +193,28 @@ void ContentVisibilityDocumentState::updateContentRelevancyForScrollIfNeeded(con
+{
+    // No need to schedule content relevancy update for first time call, since
+    // that will be handled by determineInitialVisibleContentVisibility.
+    if (m_viewportProximity.contains(element))
```
It's confusing that you have an argument which is an enum type called `viewportProximity` and a member variable called `m_viewportProximity` which is a collection of elements. I suggest you rename `m_viewportProximity`.

rwlbuis replied:
I picked a different name for the member variable.

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8446,14 +8452,17 @@ void Document::updateResizeObservations(Page& page)
+        // We need layout the whole frame tree here. Because ResizeObserver could observe element in other frame,
+        // and it could change other frame in deliverResizeObservations().
+        page.layoutIfNeeded();
+        if (m_contentVisibilityDocumentState && m_contentVisibilityDocumentState->determineInitialVisibleContentVisibility() == HadInitialVisibleContentVisibilityDetermination::Yes)
```
Why does content visibility affect this behavior? Would appreciate a comment to help me understand.

rwlbuis replied:
I added a comment.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.h:
```diff
@@ -26,14 +26,18 @@
 
 enum class DidUpdateAnyContentRelevancy : bool { No, Yes };
+// https://drafts.csswg.org/css-contain/#proximity-to-the-viewport
+enum class ViewportProximity : uint8_t { Far, Near };
```
```suggestion
enum class ViewportProximity : bool { Far, Near };
```

rwlbuis replied:
Good catch, fixed.

rwlbuis committed:
[content-visibility] Implement initial visibility determination
https://bugs.webkit.org/show_bug.cgi?id=259825

Reviewed by Tim Nguyen.

Implement initial visibility determination as specified in [1].
This replaces the custom onscreen bookkeeping with the specified
proximity to viewport concept. The updateIntersectionObservations
logic is refactored a bit so it can be used for updating any given
collection of intersection observers.

This PR also changed behaviour to not dispatch ContentVisibilityAutoStateChange
on disconnected elements, this problem is now exposed with the different timing
of initial visibility determination.

[1] https://github.com/whatwg/html/pull/9312

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-first-observation-immediate-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-first-observation-immediate.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-012-expected.txt:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::unobserve):
(WebCore::ContentVisibilityDocumentState::checkRelevancyOfContentVisibilityElement const): extract per element content relevancy check logic
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements const): bail out early if no update is specified
(WebCore::ContentVisibilityDocumentState::determineInitialVisibleContentVisibility const):
(WebCore::ContentVisibilityDocumentState::updateContentRelevancyStatusForScrollIfNeeded):
(WebCore::ContentVisibilityDocumentState::updateViewportProximity): keep track of per element viewport proximity
(WebCore::ContentVisibilityDocumentState::removeViewportProximity):
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements): Deleted.
(WebCore::ContentVisibilityDocumentState::updateOnScreenObservationTarget): Deleted.
* Source/WebCore/dom/ContentVisibilityDocumentState.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateIntersectionObservations): allow specifying the intersection observers
(WebCore::Document::updateResizeObservations): integrate according to [1]
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::contentVisibilityViewportChange): Deleted.
* Source/WebCore/dom/Element.h:

Canonical link: https://commits.webkit.org/267779@main
